src/network/netdev/macsec.c

```diff
@@ -0,0 +1,911 @@
+
+        network_config_section_free(a->section);
+
+        free(a);
+}
+
+DEFINE_NETWORK_SECTION_FUNCTIONS(TransmitAssociation, macsec_transmit_association_free);
+
+static int macsec_transmit_transmit_association_new_static(MACsec *s, const char *filename, unsigned section_line, TransmitAssociation **ret) {
+        _cleanup_(network_config_section_freep) NetworkConfigSection *n = NULL;
+        _cleanup_(macsec_transmit_association_freep) TransmitAssociation *a = NULL;
+        int r;
+
+        assert(s);
+        assert(ret);
+        assert(filename);
+        assert(section_line > 0);
+
+        r = network_config_section_new(filename, section_line, &n);
+        if (r < 0)
+                return r;
+
+        a = ordered_hashmap_get(s->transmit_association_by_section, n);
+        if (a) {
+                *ret = TAKE_PTR(a);
+                return 0;
+        }
+
+        a = new(TransmitAssociation, 1);
+        if (!a)
+                return -ENOMEM;
+
+        *a = (TransmitAssociation) {
+                .mac_sec = s,
+                .section = TAKE_PTR(n),
+        };
+
+        r = ordered_hashmap_ensure_allocated(&s->transmit_association_by_section, &network_config_hash_ops);
+        if (r < 0)
+                return r;
+
+        r = ordered_hashmap_put(s->transmit_association_by_section, a->section, a);
+        if (r < 0)
+                return r;
+
+        *ret = TAKE_PTR(a);
+
+        return 0;
+}
+
+static int netdev_macsec_fill_message(NetDev *netdev, int command, sd_netlink_message **ret) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
+        MACsec *t;
+        int r;
+
+        assert(netdev);
+        assert(netdev->ifindex > 0);
+
+        t = MACSEC(netdev);
+
+        assert(t);
+
+        r = sd_genl_message_new(netdev->manager->genl, SD_GENL_MACSEC, command, &m);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Failed to create generic netlink message: %m");
+
+        r = sd_netlink_message_append_u32(m, MACSEC_ATTR_IFINDEX, netdev->ifindex);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append MACSEC_ATTR_IFINDEX attribute: %m");
+
+        *ret = TAKE_PTR(m);
+
+        return 0;
+}
+
+static int netdev_macsec_fill_message_sci(NetDev *netdev, ReceiveChannel *c, sd_netlink_message *m) {
+        uint64_t sci;
+        MACsec *t;
+        int r;
+
+        assert(netdev);
+        assert(m);
+
+        t = MACSEC(netdev);
+
+        assert(t);
+        assert(m);
+
+        assert(c);
+        assert(c->mac);
+        assert(c->port > 0);
+
+        r = sd_netlink_message_open_container(m, MACSEC_ATTR_RXSC_CONFIG);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append MACSEC_ATTR_RXSC_CONFIG attribute: %m");
+
+        memcpy(&sci, c->mac, ETH_ALEN);
+        memcpy(((char *) &sci) + ETH_ALEN, &c->port, sizeof(c->port));
```

Comment:
hmm, is this endian safe? I presume this should be handled different on different endianess, no?

Also, wouldn't it be nicer to solve this with a uinion? i.e.

```
union {
        uint64_t as_uint64;
        struct {
                struct ether_addr addr;
                uint16_t port;
        } as_address_and_port;
} sci;
```

Or something like that?

Reply:
I tried to use it like 
```
memcpy(&sci.as_address_and_port.addr, &c->mac, ETH_ALEN);                                                                                                                          
memcpy(&sci.as_address_and_port.port, &c->port, sizeof(c->port));
```

Kernel does not accept it 

```
macsec-test: netdev could not be add receive association configurations: No such device

```

Kernel uses the same use. But we can keep union format.
https://elixir.bootlin.com/linux/latest/source/drivers/net/macsec.c#L412